src/compiler/core.ts

```diff
@@ -944,16 +944,54 @@ export function sortAndDeduplicate<T>(array: readonly T[], comparer?: Comparer<T
 /** @internal */
 export function arrayIsSorted<T>(array: readonly T[], comparer: Comparer<T>) {
     if (array.length < 2) return true;
-    let prevElement = array[0];
-    for (const element of array.slice(1)) {
-        if (comparer(prevElement, element) === Comparison.GreaterThan) {
+    for (let i = 1, len = array.length; i < len; i++) {
+        if (comparer(array[i - 1], array[i]) === Comparison.GreaterThan) {
             return false;
         }
-        prevElement = element;
     }
     return true;
 }
 
+/** @internal */
+export const enum SortKind {
+    None = 0,
+    CaseSensitive = 1 << 0,
+    CaseInsensitive = 1 << 1,
+    Both = CaseSensitive | CaseInsensitive,
+}
+
+/** @internal */
+export function detectSortCaseSensitivity(array: readonly string[], useEslintOrdering?: boolean): SortKind;
+/** @internal */
+export function detectSortCaseSensitivity<T>(array: readonly T[], useEslintOrdering: boolean, getString: (element: T) => string): SortKind;
+/** @internal */
+export function detectSortCaseSensitivity<T>(array: readonly T[], useEslintOrdering: boolean, getString?: (element: T) => string): SortKind {
+    let kind = SortKind.Both;
+    if (array.length < 2) return kind;
+    const caseSensitiveComparer = getString
+        ? (a: T, b: T) => compareStringsCaseSensitive(getString(a), getString(b))
+        : compareStringsCaseSensitive as (a: T | undefined, b: T | undefined) => Comparison;
+    const compareCaseInsensitive = useEslintOrdering ? compareStringsCaseInsensitiveEslintCompatible : compareStringsCaseInsensitive;
+    const caseInsensitiveComparer = getString
+        ? (a: T, b: T) => compareCaseInsensitive(getString(a), getString(b))
+        : compareCaseInsensitive as (a: T | undefined, b: T | undefined) => Comparison;
+    for (let i = 1, len = array.length; i < len; i++) {
+        const prevElement = array[i - 1];
+        const element = array[i];
+        if (caseSensitiveComparer(prevElement, element) === Comparison.GreaterThan) {
+            kind &= ~SortKind.CaseSensitive;
+        }
+        if (caseInsensitiveComparer(prevElement, element) === Comparison.GreaterThan) {
+            kind &= ~SortKind.CaseInsensitive;
+        }
+        if (kind === SortKind.None) {
+            return kind;
+        }

```

Comment:
I think that you can optimize this.

1. If you know that you're not case-sensitively sorted, you never have to check again; you can just always check the case-sensitive branch.
1. If you are not case-insensitively sorted, you can just bail out at that point. You won't be case-sensitively sorted.

That means the only times you ever have to do two checks for a pair of elements is the first pair that forces you to transition to only checking for case-insensitive comparison.

Reply:
Good point on (1), but (2) is not true. `['B', 'a']` is not case-insensitively sorted but is case-sensitively sorted. (That’s the entire reason this is bit flags.)

It sure sounds like the “case-insensitively sorted lists” ought to be a superset of “case-sensitively sorted lists,” doesn’t it? They’re actually a classic Venn diagram shape where there is a non-empty intersection but neither is a subset of the other. When Jake was first trying to explain the problem to me verbally, I made the same mistake.

Reply:
Ugh, why was I so convinced this was true? But yes, I see that. At least you can avoid each kind of comparison after the first failure.

Reply:
(Quiz ChatGPT about this for a good laugh)